### PR TITLE
feat: init TableMetadataBuilder

### PR DESCRIPTION
**Pull request by ZENOTME**
This PR init TableMetadataBuilder. Complete #250

### CODE REVIEW

File: crates/iceberg/src/spec/table_metadata.rs [hunk 1]
```diff
@@ -275,6 +277,78 @@ impl TableMetadata {
...
+            format_version: FormatVersion::V2,
+            table_uuid: Uuid::new_v4(),
+            location: location.ok_or_else(|| {
+                Error::new(
+                    ErrorKind::DataInvalid,
+                    "Can't create table without location",
+                )
+            })?,
+            last_sequence_number: 0,
+            last_updated_ms: 0,
```
Comment (liurenjie1024) [inline crates/iceberg/src/spec/table_metadata.rs:310]: This should be updated by current ts.

File: crates/iceberg/src/spec/table_metadata.rs [hunk 2]
```diff
@@ -275,6 +277,78 @@ impl TableMetadata {
...
+            last_column_id: 0,
```
Comment (liurenjie1024) [inline crates/iceberg/src/spec/table_metadata.rs:311]: This should be schema's largest field id.

File: crates/iceberg/src/spec/table_metadata.rs [hunk 3]
```diff
@@ -275,6 +277,78 @@ impl TableMetadata {
...
+            schemas: HashMap::from([(0, Arc::new(schema))]),
```
Comment (liurenjie1024) [inline crates/iceberg/src/spec/table_metadata.rs:312]: The key should be schema id.

File: crates/iceberg/src/spec/table_metadata.rs [hunk 4]
```diff
@@ -275,6 +277,78 @@ impl TableMetadata {
...
+            current_schema_id: 0,
```
Comment (liurenjie1024) [inline crates/iceberg/src/spec/table_metadata.rs:313]: Ditto, this should be schema's id.

File: crates/iceberg/src/spec/table_metadata.rs [hunk 5]
```diff
@@ -275,6 +277,78 @@ impl TableMetadata {
...
+            partition_specs: partition_spec
```
Comment (liurenjie1024) [inline crates/iceberg/src/spec/table_metadata.rs:314]: This is incorrect, we should bind partition spec to schema.

File: crates/iceberg/src/spec/table_metadata.rs [hunk 6]
```diff
@@ -275,6 +277,78 @@ impl TableMetadata {
...
+                })
+                .unwrap_or_default(),
+            default_spec_id: 0,
+            last_partition_id: 0,
+            properties,
+            current_snapshot_id: None,
+            snapshots: Default::default(),
+            snapshot_log: vec![],
+            metadata_log: vec![],
+            sort_orders: sort_order
```
Comment (liurenjie1024) [inline crates/iceberg/src/spec/table_metadata.rs:327]: We should bind to schema.

**Review by liurenjie1024 — COMMENTED** [review]
@ZENOTME Thanks for this pr! We need some improvements to the creation process, see [java implementation](https://github.com/apache/iceberg/blob/5e139f48505b9b622e2cf389eae2bb74d6420510/core/src/main/java/org/apache/iceberg/TableMetadata.java#L105)

We may not need to refresh ids for schema, but we need to bind partition spec and sort order to schema. We can left the binding to other pr, and forbid table creation with partition spec and sort order for now, it's up to you.

**Comment by liurenjie1024** [comment]
cc @marvinlanhenke @Fokko @Xuanwo PTAL

**Comment by ZENOTME** [comment]
>  We can left the binding to other pr, and forbid table creation with partition spec and sort order for now, it's up to you.

Let's forbid them now. I think I'm not clear with these two processes now, so let's delay them to later PR. After I complete them, I will fix this. Other fixed.

File: crates/iceberg/src/spec/partition.rs [hunk 7]
```diff
@@ -133,6 +133,25 @@ impl UnboundPartitionSpec {
     pub fn builder() -> UnboundPartitionSpecBuilder {
         UnboundPartitionSpecBuilder::default()
     }
+
+    /// Create a [`PartitionSpec`] for a new table. So it don't need
+    /// to specify the partition id for each field.
+    pub fn create_new(self) -> PartitionSpec {
```
Comment (liurenjie1024) [inline crates/iceberg/src/spec/partition.rs:139]: This is unused, show we remove this?

**Review by liurenjie1024 — COMMENTED** [review]
@ZENOTME Thanks for this pr, I think we just need to remove unused method, and others LGTM.

File: crates/iceberg/src/spec/table_metadata.rs [hunk 8]
```diff
@@ -275,6 +277,82 @@ impl TableMetadata {
...
+            metadata_log: vec![],
+            default_sort_order_id: 0,
+            refs: Default::default(),
+        };
+
+        Ok(Self(table_metadata))
+    }
+
+    /// Changes uuid of table metadata.
+    pub fn assign_uuid(mut self, uuid: Uuid) -> Result<Self> {
```
Comment (ZENOTME) [inline crates/iceberg/src/spec/table_metadata.rs:345]: If it only used in TableUpdate.apply, we can let this be `pub(crate)`.🤔

Comment (liurenjie1024) [reply crates/iceberg/src/spec/table_metadata.rs]: Currently it's only used here, but it's also intended for user to construct table metadata, so I think it's ok to leave it `pub`

File: crates/iceberg/src/spec/table_metadata.rs [hunk 9]
```diff
@@ -275,6 +277,82 @@ impl TableMetadata {
...
+pub struct TableMetadataBuilder(TableMetadata);
+
+impl TableMetadataBuilder {
+    /// Creates a new table metadata builder from the given table metadata.
+    pub fn new(origin: TableMetadata) -> Self {
+        Self(origin)
+    }
+
+    /// Creates a new table metadata builder from the given table creation.
+    pub fn from_table_creation(table_creation: TableCreation) -> Result<Self> {
```
Comment (marvinlanhenke) [inline crates/iceberg/src/spec/table_metadata.rs:290]: I really like this; makes me life a lot easier in catalog/hms

**Review by liurenjie1024 — APPROVED** [review]
LGTM,Thanks!

**Pull request merged**